src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -200,12 +200,14 @@ async Task EnsureLong()
                 // We have a special case when reading async notifications - a timeout may be normal
                 // shouldn't be fatal
                 // Note that mono throws SocketException with the wrong error (see #1330)
-                catch (IOException e) when (
-                    dontBreakOnTimeouts && (e.InnerException as SocketException)?.SocketErrorCode ==
-                       (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock)
-                )
+                catch (IOException e) when ((e.InnerException as SocketException)?.SocketErrorCode ==
+                       (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock))
                 {
-                    throw new TimeoutException("Timeout while reading from stream");
+                    if (dontBreakOnTimeouts)
+                        throw new TimeoutException("Timeout while reading from stream");
+                    else
+                        throw Connector.Break(new NpgsqlException("Exception while reading from stream",
+                            new TimeoutException("Timeout during reading attempt", e)));
```

Comment:
Do we think there's any value in including the IOException as an inner exception of the TimeoutException? I don't think it adds any information beyond what the TimeoutException already conveys, and it will make the user-visible ToString much more verbose and hard to read.